### PR TITLE
Avoid NaNs in SimClustering

### PR DESCRIPTION
**Pull request by kpedro88**
Some Phase 2 TDR production workflows encountered SimClusters with NaN fractions for certain hits.

I found the likely culprit in `CaloTruthAccumulator` that could make NaN values, and added a protection. I also changed the exception in `Cluster3DPCACalculator` to a warning, so the jobs can still run in case there's another problem that I missed. (These NaNs happen very rarely and are almost impossible to reproduce given we're running 200PU w/ 4 threads.)

This PR will be backported to 91X.

attn: @lgray

### CODE REVIEW

**Comment by kpedro88** [comment]
assign upgrade

**Comment by kpedro88** [comment]
please test

**Comment by civanch** [comment]
@kpedro88 , if there is a rare problem I would extend maximally warning printout dumping max information about such cluster.

**Comment by kpedro88** [comment]
@civanch the map `m_detIdToTotalSimEnergy` is filled numerous times (whenever SimHits are added, i.e. signal and pileup), and then it's used in a different function at the end of the processing. This makes it hard to get a lot of information when the problem is actually noticed. However, I realized I wasn't printing out the actual DetId, so I updated the warning to at least do that.

**Comment by kpedro88** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
@slava77 @perrotta please sign
@davidlange6 this is needed very urgently for TDR production

File: SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc [hunk 1]
```diff
@@ -426,13 +430,7 @@ template<class T> void CaloTruthAccumulator::fillSimHits( std::vector<std::pair<
       DetId id(0);
       const uint32_t simId = simHit.id();
       if( isHcal ) {
-	int subdet, z, depth0, eta0, phi0, lay;
-	HcalTestNumbering::unpackHcalIndex(simId, subdet, z, depth0, eta0, phi0, lay);
-	int sign = (z==0) ? (-1):(1);
-	HcalDDDRecConstants::HcalID tempid = hcddd_->getHCID(subdet, sign*eta0, phi0, lay, depth0);
-	if (subdet==int(HcalEndcap)) {
-	  id = HcalDetId(HcalEndcap,sign*tempid.eta,tempid.phi,tempid.depth);
```
Comment (perrotta) [inline SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc]: Out of curiosity: the fact that before "id" was assigned only for HcalEndcap (while now it is for any Hcal) was a kind of bug also fixed with this PR?

Comment (kpedro88) [inline SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc]: Actually no, this is a bug in the change I made, will fix

**Comment by kpedro88** [comment]
please test

**Comment by perrotta** [comment]
+1
- Rather technical: the code reflects what it is intended to do
- No difference expected in standard outputs (only once in a while on a very large stats, according to the description notes) and none is observed in jenkins outputs

**Comment by kpedro88** [comment]
+1

**Comment by davidlange6** [comment]
merge

**Comment by slava77** [comment]
> Some Phase 2 TDR production workflows encountered SimClusters with NaN fractions for certain hits.

@kpedro88 please clarify which workflows have this problem, perhaps also add a link to a discussion thread in HN
Thank you.

**Comment by kpedro88** [comment]
@slava77 see https://hypernews.cern.ch/HyperNews/CMS/get/prep-ops/4147.html

**Comment by slava77** [comment]
On 8/3/17 6:10 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> see
> https://hypernews.cern.ch/HyperNews/CMS/get/prep-ops/4147.html

please make a github issue to resolve this eventually.
Is it understood how we get the NaNs at least in this case?
The best decision would be to eliminate NaN creation at the origin
rather than filter out in every  code consuming the products with NaNs.

**Comment by kpedro88** [comment]
@slava77 this PR attempts to do both - I identified the only place where I thought NaNs could be produced in `CaloTruthAccumulator` and added a protection. Skipping any affected hits in PF is just a patch to make sure the TDR production can continue.

It's very hard to reproduce the problem - these were 200PU workflows running on 4 threads, so I don't think we can get the exact DIGI step for an affected event. I'll open an issue for further investigation.

**Pull request merged**